hledger-lib/Hledger/Read/CsvReader.hs

```diff
@@ -899,98 +864,98 @@ transactionFromCsvRecord sourcepos rules record = t
--- The transaction which will contain this posting is also provided,
--- so we can build the usual transaction<->posting cyclic reference.
-mkPosting ::
-  CsvRules -> CsvRecord -> String ->
-  HledgerFieldName -> HledgerFieldName -> HledgerFieldName ->
-  HledgerFieldName -> HledgerFieldName -> HledgerFieldName ->
-  Transaction ->
-  Maybe (Posting, Bool)
-mkPosting rules record number accountFld amountFld amountInFld amountOutFld balanceFld commentFld t =
-  -- if we have figured out an account N, make a posting N
-  case maccountAndIsFinal of
-    Nothing            -> Nothing
-    Just (acct, final) ->
-      Just (posting{paccount          = accountNameWithoutPostingType acct
-                   ,pamount           = fromMaybe missingmixedamt mamount
-                   ,ptransaction      = Just t
-                   ,pbalanceassertion = mkBalanceAssertion rules record <$> mbalance
-                   ,pcomment          = comment
-                   ,ptype             = accountNamePostingType acct}
-           ,final)
+-- -- | Given CSV rules and a CSV record, generate the corresponding transaction's
+-- -- Nth posting, if sufficient fields have been assigned for it.
+-- -- N is provided as a string.
+-- -- The names of the required fields are provided, allowing more flexibility.
+-- -- The transaction which will contain this posting is also provided,
+-- -- so we can build the usual transaction<->posting cyclic reference.
+-- mkPosting :: CsvRules -> CsvRecord -> String -> Transaction -> Maybe (Posting, Bool)
+-- mkPosting rules record n t =
+
+-- | Figure out the amount specified for posting N, if any.
+-- Looks for a non-zero amount assigned to one of "amountN", "amountN-in", "amountN-out".
+-- Postings 1 or 2 also look at "amount", "amount-in", "amount-out".
+-- Throws an error if more than one of these has a non-zero amount assigned.
+-- A currency symbol to prepend to the amount, if any, is provided,
+-- and whether posting 1 requires balancing or not.
+getAmount :: CsvRules -> CsvRecord -> String -> Bool -> Int -> Maybe MixedAmount
+getAmount rules record currency p1IsVirtual n =
+  let
+    unnumberedfieldnames = ["amount","amount-in","amount-out"]
+    fieldnames = map (("amount"++show n)++) ["","-in","-out"]
+                 -- For posting 1, also recognise the old field names.
+                 -- For posting 2, the same but only if posting 1 needs balancing.
+                 ++ if n==1 || n==2 && not p1IsVirtual then unnumberedfieldnames else []
+    nonzeroamounts = [(f,a') | f <- fieldnames
+                     , Just v@(_:_) <- [strip . renderTemplate rules record <$> hledgerField rules record f]
+                     , let a = parseAmount rules record currency v
+                     , not $ isZeroMixedAmount a
+                       -- With an old field name, flip the sign in posting 2
+                     , let a' = if f `elem` unnumberedfieldnames && n==2 then -a else a
```

Comment:
I think that for better compatibility with pre-1.17 this line should be:
```
let a' = if f `elem` unnumberedfieldnames && n==2 then costOfMixedAmount(-a) else a

```

Reply:
Thanks!